File: docs/chapters/01-solid-and-core-engineering-principles/go/interface_segregation_repository.go
// Scenario: billing dashboard reads invoices; settlement workers write ledger rows
// Demonstrates: Interface Segregation — Reader vs Writer interfaces in Go
// Trade-off: compile-time enforcement via small interfaces; no fat Repository

package main

type Invoice struct {
	ID          string
	AmountCents int64
}

type LedgerEntry struct {
	ID        string
	InvoiceID string
}

type InvoiceReader interface {
	FindByID(id string) (Invoice, bool)
	ListOpen() []Invoice
}

type LedgerWriter interface {
	Append(entry LedgerEntry) error
}

type ReportingService struct{ invoices InvoiceReader }

func (s *ReportingService) OpenBalance() int64 {
	var sum int64
	for _, inv := range s.invoices.ListOpen() {
		sum += inv.AmountCents
	}
	return sum
}

type SettlementWorker struct{ ledger LedgerWriter }

func (w *SettlementWorker) Settle(inv Invoice) error {
	return w.ledger.Append(LedgerEntry{ID: "le-" + inv.ID, InvoiceID: inv.ID})
}

File: docs/chapters/01-solid-and-core-engineering-principles/java/InterfaceSegregationRepository.java
// Scenario: billing dashboard reads invoices; settlement workers write ledger rows
// Demonstrates: Interface Segregation — narrow Reader vs Writer ports
// Trade-off: two interfaces vs fat repo with unsupported write methods

import java.util.List;
import java.util.Optional;

record Invoice(String id, long amountCents) {}
record LedgerEntry(String id, String invoiceId) {}

interface InvoiceReader {
    Optional<Invoice> findById(String id);
    List<Invoice> listOpen();
}

interface LedgerWriter {
    void append(LedgerEntry entry);
}

final class ReportingService {
    private final InvoiceReader invoices;
    ReportingService(InvoiceReader invoices) { this.invoices = invoices; }
    long openBalance() {
        return invoices.listOpen().stream().mapToLong(Invoice::amountCents).sum();
    }
}

final class SettlementWorker {
    private final LedgerWriter ledger;
    SettlementWorker(LedgerWriter ledger) { this.ledger = ledger; }
    void settle(Invoice inv) {
        ledger.append(new LedgerEntry("le-" + inv.id(), inv.id()));
    }
}
